docs/TICCL-mergelex.1
.TH TICCL-mergelex 1 "2020 mar 09"

.SH NAME
TICCL-mergelex - a tool to merge several lexicons. Note this tool does not form part of the TICCL pipeline proper!

.SH SYNOPSIS

TICCL-mergelex [options] FILES/DIR

.SH DESCRIPTION
.B TICC-mergelex
will create a merged lexicon from a range of lexicon files in
.B TICCL-stats
or
.B FoLiA-stats
format.
The output will be a 2 or 4 columned tab separated file, extension: *tsv
(4 columns when -p is specified)

.SH OPTIONS
.B -p
.RS
output percentages too.
.RE

.B -e
expr
.RS
specifies the expression all input files should match with.
.RE

.B -R
.RS
Recursively search the given directory
.RE

.B -o
output_file
.RS
Write results to this file.
.RE

.B -t
or
.B --threads
num_threads
.RS
use 'num_threads' threads to run on. You may us --threads="max" to use as many
threads as possible. This will allocate 2 processors less than given by the
$OMP_NUM_THREADS environment variable, leaving some processor power for other
purposes.
.RE

.B -v
.RS
be more verbose
.RE

.B -V
or
.B --version
.RS
Show version
.RE

.B -h
or
.B --help
.RS
usage info
.RE

.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot lamasoftware@science.ru.nl

Martin Reynaert reynaert@uvt.nl

.SH SEE ALSO
.BR TICCL-lexstat (1)
.BR FoLiA-stats (1)


NEWS
ticcltools 0.6 2018-06-05
[Ko vander Sloot]
Intermediate release, with a lot of new code to handle N-grams
Also a lot of refactoring is done, for more clear and maintainable code.
This is work in progress still.

* TICCL-unk:
  - more extensive acronym detection
  - fixed artifreq problems in 'clean' punctuated words
  - added filters for 'unwanted' characters
  - added a ligature filter to convert evil ligatures
  - normalize all hyphens to a 'normal' one (-)
  - use a better definition of punctuation (unicode character class is not
    good enough to decide)

* TICCL-lexstat:
  - the 'separator' symbol should get freq=0, so it isn't counted
  - the clip value is added to the output filename

* TICCL-indexer:
  - indexer and indexerNT now produce the same output, using different
    strategies when a --foci files is used.

* TICCL-LDcalc:
  major overhaul for n-grams
  - added a ngram point column to the output (so NOT backward compatible!)
  - produce a '.short' list for short word corrections
  - produce a '.ambi' file with a list of n-grams related to short words
  - prune a lot of ngrams from the output

* TICCL-rank:
 - output is sorted now
 - honor the ngram-points from the new LDcalc. (so NOT backward compatible!)

* TICCL-chain: new module to chain ranked files

* TICCL-lexclean:
  -added a -x option for 'inverse' alphabet

* TICCL-anahash:
  - added a --list option to produce a list of words and anagram values

[Maarten van Gompel]
* added metadata file: codemeta.json

ticcltools 0.5 2018-02-19
[Ko van der Sloot]
* updated configuration. also for Mac OSX
* use of more ticcutils stuff: diacritics filter
* added a TICCL-mergelex program
* the OMP_THREAD_LIMIT environment variable was ignored sometimes
* TICCL-unk:
   - fixed a problem in artifreq handling
   - changed acronym detection (work in  progress)
   - added -o option
  TICCL-lexstat:
   - added TTR output
   - added -o option
  TICCL-indexer
  - now also handles --foci file. with some speed-up
  - added a -t option
  TICCL-LDcalc:
  - be less picky on a few wrong lines in the data
* added some tests
* when libroaring is installed we built roaring versions of some modules (experimental)
* updated man pages

ticcltools 0.4 2017-04-04
[Ko van der Sloot]
* first official release.
  - added functions to test on Word2Vec datafiles
  - refactoring and modernizing stuff all around

ticcltools 0.3 2016-01-21
* upload ready
ticcltools 0.2 2016-01-14
[Ko van der Sloot]
* repository moved to GitHub
* Travis support
* first 'distributable' version
* added TICCL-stats program
* added W2V-near and W2V-dist programs bases on new libticcl lib


[0.1] Ko vd Sloot
 started autoconfiscating


src/TICCL-mergelex.cxx
/*
  Copyright (c) 2006 - 2020
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of ticcltools

  foliatools is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  foliatools is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/frog/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl

*/

#include <cmath>
#include <string>
#include <map>
#include <vector>
#include <iostream>
#include <fstream>

#include "ticcutils/CommandLine.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/StringOps.h"
#include "ticcutils/XMLtools.h"
#include "ticcutils/Unicode.h"

#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif

using namespace	std;
using namespace	TiCC;

bool verbose = false;

void create_wf_list( const map<string, unsigned int>& wc,
		     const string& filename, unsigned int total_in, bool doperc ){
  ofstream os( filename );
  if ( !os ){
    cerr << "failed to create outputfile '" << filename << "'" << endl;
    exit(EXIT_FAILURE);
  }
  map<unsigned int, set<string> > wf;
  for( const auto& cit : wc ){
    wf[cit.second].insert( cit.first );
  }
  unsigned int sum=0;
  unsigned int types=0;
  map<unsigned int, set<string> >::const_reverse_iterator wit = wf.rbegin();
  while ( wit != wf.rend() ){
    for( const auto& sit : wit->second ){
      sum += wit->first;
      os << sit << "\t" << wit->first;
      if ( doperc ){
	os << "\t" << sum << "\t" << 100 * double(sum)/total_in;
      }
      os << endl;
      ++types;
    }
    ++wit;
  }
#pragma omp critical
  {
    cout << "created WordFreq list '" << filename << "'" << endl
	 << "with " << total_in << " tokens and " << types
	 << " types. TTR= " << (double)types/total_in
	 << ", the angle is " << atan((double)types/total_in)*180/M_PI
	 << " degrees" << endl;
  }
}

size_t read_words( const string& doc_name, map<string,unsigned int>& wc ){
  size_t word_total = 0;
  ifstream is( doc_name );
  string line;
  while ( getline( is, line ) ){
    vector<string> v;
    if ( TiCC::split( line, v ) < 2 ){
      cerr << "invalid input: " << line << endl;
      continue;
    }
    string wrd = v[0];
    size_t frq = stringTo<int>(v[1]);
#pragma omp critical
    {
      wc[wrd] += frq;
    }
    word_total += frq;
  }
  return word_total;
}


void usage( const string& name ){
  cerr << "Usage: " << name << " [options] file/dir" << endl;
  cerr << "\t TICCL-mergelex will create a merged lexicon from a range of" << endl;
  cerr << "\t\t lexicon files in TICCL-lexstat or FoLiA-stats format." << endl;
  cerr << "\t\t The output will be a 2 or 4 columned tab separated file, extension: .tsv " << endl;
  cerr << "\t\t (4 columns when -p is specified)" << endl;
  cerr << "\t-p\t output percentages too. " << endl;
  cerr << "\t-e expr:\t specify the expression all input files should match with." << endl;
  cerr << "\t-o\t name of the output file(s) prefix." << endl;
  cerr << "\t-R\t search the dirs recursively (when appropriate)." << endl;
  cerr << "\t-t <threads> or --threads <threads> Number of threads to run on." << endl;
  cerr << "\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t reasonable value. (OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-v\t very verbose output." << endl;
  cerr << "\t-h or --help\t this message" << endl;
  cerr << "\t-V or --version\t show version " << endl;
}

int main( int argc, char *argv[] ){
  CL_Options opts( "hVve:t:o:Rp", "threads:,help,version" );
  try {
    opts.init(argc,argv);
  }
  catch( OptionError& e ){
    cerr << e.what() << endl;
    usage(argv[0]);
    exit( EXIT_FAILURE );
  }
  string progname = opts.prog_name();
  if ( argc < 2 ){
    usage( progname );
    exit(EXIT_FAILURE);
  }
  string expression;
  string out_prefix;
  if ( opts.extract('V' ) ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract('h' ) ){
    usage(progname);
    exit(EXIT_SUCCESS);
  }
  verbose = opts.extract( 'v' );
  bool recursiveDirs = opts.extract( 'R' );
  if ( !opts.extract( 'o', out_prefix ) ){
    cerr << "an output filename prefix is required. (-o option) " << endl;
    exit(EXIT_FAILURE);
  }
  string value = "1";
  if ( !opts.extract( 't', value ) ){
    opts.extract( "threads", value );
  }
#ifdef HAVE_OPENMP
  int numThreads=1;
  if ( TiCC::lowercase(value) == "max" ){
    numThreads = omp_get_max_threads() - 2;
    omp_set_num_threads( numThreads );
    cout << "running on " << numThreads << " threads." << endl;
  }
  else {
    if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
    omp_set_num_threads( numThreads );
    cout << "running on " << numThreads << " threads." << endl;
  }
#else
  if ( value != "1" ){
    cerr << "unable to set number of threads!.\nNo OpenMP support available!"
	 <<endl;
    exit(EXIT_FAILURE);
  }
#endif

  opts.extract('e', expression );
  bool dopercentage = opts.extract('p');
  if ( !opts.empty() ){
    cerr << "unsupported options : " << opts.toString() << endl;
    usage(progname);
    exit(EXIT_FAILURE);
  }

  vector<string> mass_opts = opts.getMassOpts();
  if ( mass_opts.empty() ){
    cerr << "no file or dir specified!" << endl;
    exit(EXIT_FAILURE);
  }
  vector<string> file_names;
  string dir_name;
  if ( mass_opts.size() > 1 ){
    // assume a list of files
    file_names = mass_opts;
  }
  else {
    dir_name = mass_opts[0];
    try {
      file_names = searchFilesMatch( dir_name, expression, recursiveDirs );
    }
    catch ( const exception& e ){
      cerr << e.what() << endl;
      exit(EXIT_FAILURE);
    }
  }
  size_t to_do = file_names.size();
  if ( to_do == 0 ){
    cerr << "no matching files found" << endl;
    exit(EXIT_SUCCESS);
  }

  string::size_type pos = out_prefix.find( "." );
  if ( pos != string::npos && pos == out_prefix.length()-1 ){
    // outputname ends with a .
    out_prefix = out_prefix.substr(0,pos);
  }
  pos = out_prefix.find( "/" );
  if ( pos != string::npos && pos == out_prefix.length()-1 ){
    // outputname ends with a /
    out_prefix += "ticclstats";
  }

  if ( to_do > 1 ){
    cout << "start processing of " << to_do << " files " << endl;
  }
  map<string,unsigned int> wc;
  unsigned int word_total =0;
#pragma omp parallel for shared(file_names,word_total,wc)
  for ( size_t fn=0; fn < file_names.size(); ++fn ){
    string doc_name = file_names[fn];
    unsigned int word_count = read_words( doc_name, wc );
    word_total += word_count;
#pragma omp critical
    {
      cout << "Processed :" << doc_name << " with " << word_count << " words,"
	   << " still " << --to_do << " files to go." << endl;
    }
  }
  if ( !dir_name.empty() ){
    cout << "done processsing directory '" << dir_name << "' in total "
	 << word_total << " words were found." << endl;
  }
  cout << "start outputting the results" << endl;
  string file_name = out_prefix + ".wordfreqlist.tsv";
  create_wf_list( wc, file_name, word_total, dopercentage );
  exit( EXIT_SUCCESS );
}
